### PR TITLE
[flang1] Skip deep deallocation for allocatable temporary variable

### PR DESCRIPTION
**Pull request by Narutoworld**
Flang skips deep allocation for a compiler-generated temporary variable,
but does not check this condition for its deallocation counterpart.
This fix adds a condition check to address this problem.

### CODE REVIEW

**Comment by Narutoworld** [comment]
From `transfrm.c`, the deep allocate for temp is skipped by `!HCCSYMG(sptrdest)`.
```
  if (!ALLOCATTRG(sptrdest) || A_TYPEG(astdest) == A_SUBSCR) {
    if (DTYG(dtypedest) == TY_DERIVED && !HCCSYMG(sptrdest) && !XBIT(54, 0x4) &&
        allocatable_member(sptrdest)) {
      handle_allocatable_members(astdest, astsrc, std, false);
      ast_to_comment(astasgn);
      return;
    }
```

**Comment by bryanpkc** [comment]
Please add a test case or two.

File: tools/flang1/flang1exe/func.c [hunk 1]
```diff
@@ -3967,13 +3967,24 @@ rewrite_calls(void)
...
-             * allocatable nor pointer
-             */
-            nop_dealloc(sptr_lhs, ast);
+          if (!HCCSYMG(sptr_lhs)) {
+            rewrite_deallocate(A_SRCG(ast), false, std);
+            if (!ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) &&
+                !POINTERG(sptr_lhs)) {
+              /* Has allocatable members but item itself is not
+              * allocatable nor pointer
+              */
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/func.c:3976]: These two lines need one more space before the indented comment.

Comment (bryanpkc) [reply tools/flang1/flang1exe/func.c:3976]: Same thing below.

File: tools/flang1/flang1exe/func.c [hunk 2]
```diff
@@ -3967,13 +3967,24 @@ rewrite_calls(void)
...
+              nop_dealloc(sptr_lhs, ast);
+            }
+          } else if (HCCSYMG(sptr_lhs) && !ALLOCG(sptr_lhs) &&
+            !ALLOCATTRG(sptr_lhs) && !POINTERG(sptr_lhs)) {
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/func.c:3980]: Please indent this line so that `!ALLOCATTRG` is aligned with `HCCSYMG` above.

File: tools/flang1/flang1exe/func.c [hunk 3]
```diff
@@ -3967,13 +3967,24 @@ rewrite_calls(void)
...
+              /* if sptr_lhs is a temp variable, but it has
+              * allocatable members and item itself is not
+              * allocatable nor pointer.
+              * still rewrite deallocate and perform nop_dealloc
+              */
+              rewrite_deallocate(A_SRCG(ast), false, std);
+              nop_dealloc(sptr_lhs, ast);
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/func.c:3987]: Expanding this block into two cases minimizes changes to the original code, but makes the logic unclear. I suggest checking the allocatable/pointer property only once, and simplify the control flow to call `rewrite_deallocate` and `nop_dealloc` only once each.

Is a compiler-generated temporary allocatable variable guaranteed to be shallow-copied to another variable? If it is, please cite the relevant code in the comments. Otherwise, this code is at risk of leaking memory.

File: tools/flang1/flang1exe/func.c [hunk 4]
```diff
@@ -1975,8 +1975,8 @@ rewrite_func_ast(int func_ast, int func_args, int lhs)
        list. Those two integer kind for the exitstat and cmdstat argument
        respectively.
      */
-    ARGT_ARG(newargt, nargs - 2) = mk_cval(size_of(stb.user.dt_int), DT_INT4);    
-    ARGT_ARG(newargt, nargs - 1) = mk_cval(size_of(stb.user.dt_int), DT_INT4);
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/func.c:1979]: Please avoid formatting code that is not relevant to your bug-fix.

File: tools/flang1/flang1exe/func.c (hunk 3 above)
Comment (bryanpkc) [reply tools/flang1/flang1exe/func.c:3987]: I understand that you have cited the code in transfrm.c that transforms allocatable assignment, but it seems that there is no way for the deallocation transformation to know that the temporary variable has been shallow-copied (or in other words, that its elements have escaped to another container).

Also, does your fix need to check the state of `XBIT(54, 0x4)`?

Comment (Narutoworld) [reply tools/flang1/flang1exe/func.c:3987]: 1. Thanks for your suggesion, I simplified it based on your comment.
2. I think compiler-generated temporary allocatable variables are guaranteed to be shallow-copied. Here is the reason.

- The shallow-copied operation is performed by `tranform()` function.
- It make a call to `find_allocatable_assignment()`.
- Inside `find_allocatable_assignment()`, `rewrite_allocatable_assignment()` is responsible for rewriting.
- And `handle_allocatable_members()` actually rewrites the ast nodes.
  ```
if (!ALLOCATTRG(sptrdest) || A_TYPEG(astdest) == A_SUBSCR) {
  if (DTYG(dtypedest) == TY_DERIVED && !HCCSYMG(sptrdest) && !XBIT(54, 0x4) &&
      allocatable_member(sptrdest) &&
      /* Stop calling handle_allocatable_members() if the member is an
       * allocatable component of recursive type, otherwise would cause an
       * infinite loop. */
      !handle_recursive_alloc_members) {
    handle_allocatable_members(astdest, astsrc, std, false);
    ast_to_comment(astasgn);
    return;
  }
  if (STYPEG(sptrdest) == ST_MEMBER && !XBIT(54, 0x4) && XBIT(54, 0x1)) {
    /* FS#19118 - this typically occurs with an intrinsic assignment
     * that has a structure constructor on the right hand side. We need
     * to make sure the parent object is allocated when -Mallocatable=03
     * is used.
     */
    astdest = A_PARENTG(astdest);
    if (A_TYPEG(astdest) == A_SUBSCR)
      astdest = A_LOPG(astdest);
    if (A_TYPEG(astdest) == A_MEM) {
      sptrdest = A_SPTRG(A_MEMG(astdest));
    } else
      sptrdest = A_SPTRG(astdest);
    dtypedest = A_DTYPEG(astdest);
    if (!ALLOCATTRG(sptrdest) || DTY(dtypedest) == TY_ARRAY)
      return;
    alloc_scalar_parent_only = TRUE; /* not returning on this one path */
  } else {
    return;
  }
}
```

- The above code snippet contains the first occurrence of `handle_allocatable_members()`.
- Looking closely, anything satisfied `!ALLOCATTRG(sptrdest)` shall fall into this condition.
- According to documentation, only explicitly declared allocatable attribute would has this flag.
- Compiler created temporaries should not have **ALLOCATTR** flag and will fall into this category.

Comment (Narutoworld) [reply tools/flang1/flang1exe/func.c:3987]: 3. Rewriting deallocation happens before rewriting allocattion. And it seems no code is used to keep track of temporaries.
4. I don't think there are needed since those xflag are used for allocation part. Not included in the original deallocation code either.

Comment (bryanpkc) [reply tools/flang1/flang1exe/func.c:3987]: Thank you very much for the detailed explanation. Can you summarize your reasoning (in not as many words) in the code comments, or the commit message?

File: tools/flang1/flang1exe/func.c (hunk 4 above)
Comment (bryanpkc) [reply tools/flang1/flang1exe/func.c:1979]: The current branch still makes formatting changes in irrelevant parts of the file.

File: tools/flang1/flang1exe/func.c (hunk 3 above)
Comment (bryanpkc) [reply tools/flang1/flang1exe/func.c:3987]: > Thanks for your suggesion, I simplified it based on your comment.

Your current branch is still the same as before.

**Comment by Narutoworld** [comment]
A testcase is uploaded

File: tools/flang1/flang1exe/func.c [hunk 5]
```diff
@@ -3967,14 +3967,22 @@ rewrite_calls(void)
...
         if (allocatable_member(sptr_lhs)) {
-          rewrite_deallocate(A_SRCG(ast), false, std);
-          if (!ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) &&
-              !POINTERG(sptr_lhs)) {
-            /* Has allocatable members but item itself is not
-             * allocatable nor pointer
-             */
+          /* Has allocatable members but item itself is not
+          * allocatable nor pointer
+          */
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/func.c:3972]: For a comment block, align the asterisks in the same column. See the original code for example.

File: tools/flang1/flang1exe/func.c [hunk 6]
```diff
@@ -3967,14 +3967,22 @@ rewrite_calls(void)
...
+          bool lhs_not_allocatable =
+            !ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) && !POINTERG(sptr_lhs);
+          /* HCCSYMG() check if sptr_lhs is a temp variable or not. 
+          * It is used inside rewrite_allocatable_assignment() fuction 
+          * to guard handle_allocatable_members() which perform rewriting.
+          * As as result, compiler-generated temporary allocatable variables will
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/func.c:3978]: `perform` &rarr; `performs`

`As as result` &rarr; `As a result`

File: tools/flang1/flang1exe/func.c [hunk 7]
```diff
@@ -3967,14 +3967,22 @@ rewrite_calls(void)
...
+           * allocatable nor pointer
+           */
+          bool lhs_not_allocatable =
+            !ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) && !POINTERG(sptr_lhs);
+          /* HCCSYMG() checkes if sptr_lhs is a temp variable or not. 
+           * It is used inside rewrite_allocatable_assignment() fuction 
+           * to guard handle_allocatable_members() which performs rewriting.
+           * As a result, compiler-generated temporary allocatable variables will
+           * only perform shallow copy.
+           * Adding this condition to skip deep deallocation for temp variables.
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/func.c:3980]:
```
           * As a result, compiler-generated temporary allocatable variables will
           * only undergo shallow copying. HCCSYMG() is used here to prevent deep
           * deallocation for temporary allocatable variables.
```

File: test/f90_correct/inc/remove_deep_deallocatable.mk [hunk 8]
```diff
@@ -0,0 +1,25 @@
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+
+########## Make rule for test remove_deep_deallocatable   ########
```
Comment (bryanpkc) [inline test/f90_correct/inc/remove_deep_deallocatable.mk:7]: Most tests don't have this line, I think. Let's try to reduce meaningless boilerplate.

File: test/f90_correct/inc/remove_deep_deallocatable.mk [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+
+remove_deep_deallocatable: run
```
Comment (bryanpkc) [inline test/f90_correct/inc/remove_deep_deallocatable.mk:10]:
```
$(TEST): run
```

Comment (bryanpkc) [reply test/f90_correct/inc/remove_deep_deallocatable.mk:10]: Occurrences of `remove_deep_deallocate` in this file should be replaced with `$(TEST)`.

File: tools/flang1/flang1exe/func.c [hunk 10]
```diff
@@ -3967,14 +3967,22 @@ rewrite_calls(void)
...
-              !POINTERG(sptr_lhs)) {
-            /* Has allocatable members but item itself is not
-             * allocatable nor pointer
-             */
+          /* Has allocatable members but item itself is not
+           * allocatable nor pointer
+           */
+          bool lhs_not_allocatable =
+            !ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) && !POINTERG(sptr_lhs);
+          /* HCCSYMG() checkes if sptr_lhs is a temp variable or not.
```
Comment (mleair) [inline tools/flang1/flang1exe/func.c:3975]: checkes: Misspelled "checks"

**Comment by Narutoworld** [comment]
Add another testcase

File: test/f90_correct/src/remove_deep_deallocatable.f08 [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    allocate(arg1(1)%r, source=[99,199,1999])
+    allocate(arg2(1))
+    allocate(arg2(1)%r, source=[42,142])
+    d = [arg1, arg2]
+    deallocate(arg1)
+    deallocate(arg2)
+
+    call check(d(1)%r, [99,199,1999], 3)
+    call check(d(2)%r, [42,142], 2)
+  end program test
```
Comment (bryanpkc) [inline test/f90_correct/src/remove_deep_deallocatable.f08:26]: NIT: Unnecessary indentation by 2 spaces.

**Comment by Narutoworld** [comment]
Commit message revised

**Review by mleair — APPROVED** [review]
LGTM. Thanks for the fix and the nice test case.

File: tools/flang1/flang1exe/func.c [hunk 12]
```diff
@@ -3967,14 +3967,22 @@ rewrite_calls(void)
...
-            /* Has allocatable members but item itself is not
-             * allocatable nor pointer
-             */
+          /* Has allocatable members but item itself is not
+           * allocatable nor pointer
+           */
+          bool lhs_not_allocatable =
+            !ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) && !POINTERG(sptr_lhs);
+          /* HCCSYMG() checks if sptr_lhs is a temp variable or not. 
+           * It is used inside rewrite_allocatable_assignment() fuction
```
Comment (mleair) [inline tools/flang1/flang1exe/func.c:3976]: Ooops! Just spotted another typo... function not fuction.  ;-)

**Comment by bryanpkc** [comment]
We need to run a spellchecker on our patches... 😆

File: tools/flang1/flang1exe/func.c (hunk 12 above)
Comment (bryanpkc) [reply tools/flang1/flang1exe/func.c:3976]: We need to run a spellchecker on our patches... 😆

**Comment by bryanpkc** [comment]
gentle ping @pawosm-arm @shivaramaarao

**Review by shivaramaarao — APPROVED** [review]
